### PR TITLE
add types to slice

### PR DESCRIPTION
**Pull request by wontruefree**
I added some types to slice and renamed the `hexstring` function that takes a pointer to `to_unsafe_hexstring`.

### CODE REVIEW

**Comment by konovod** [comment]
@wontruefree `hexstring` doesn't take only pointer, it can take a Slice.

File: src/slice.cr [hunk 1]
```diff
@@ -493,11 +495,11 @@ struct Slice(T)
...
   end
 
-  private def to_hex(c)
+  private def to_hex(c : Int) : UInt8
     ((c < 10 ? 48_u8 : 87_u8) + c)
   end
 
-  def bytesize
+  def bytesize : Int32
     sizeof(T) * size
```
Comment (yxhuvud) [inline src/slice.cr]: Whoa, this may be a candidate for being upscaled to Int64.

Comment (wontruefree) [reply src/slice.cr]: or maybe this could just be `Int`

**Comment by wontruefree** [comment]
@konovod It also mutates the values in place though.  This seems like an unsafe action to me.
Also, I don't mind leaving it untyped until the API gets more specific.

File: src/slice.cr (hunk 1 above)
Comment (yxhuvud) [reply src/slice.cr]: The problem is with the implementation, not the type. It will be wrong if sizeof(T) * size > Int32::MAX.

Comment (wontruefree) [reply src/slice.cr]: That makes sense but I would want to pull that into another PR

**Pull request closed without merging**